torch/_dynamo/variables/builder.py

```diff
@@ -1166,18 +1166,29 @@ def wrap_to_fake_tensor_and_record(
         if tx.fake_mode.shape_env is not None:
             dynamic_dims = []
             constraint_dims = []
+            automatic_marked_dims = set()
+            if isinstance(source, LocalSource):
+                dynamic_plan = tx.output.dynamic_plan
+                local_name = source.local_name
+                if dynamic_plan and local_name in dynamic_plan:
+                    automatic_marked_dims = dynamic_plan[local_name]
+
             for i in range(e.dim()):
                 # NB: mark dynamic has precedence over static
                 marked_dynamic = i in getattr(e, "_dynamo_dynamic_indices", set())
                 marked_static = i in getattr(e, "_dynamo_static_indices", set())
+                # NB: Any user markings have precedence over automatic
+                automatic_dynamic = i in automatic_marked_dims
```

Comment:
This is not implemented correctly. If a user marked a dimension as static, but we then inferred it's dynamic, the constraint processing will generate a non-None constraint, and then we will always choose DimDynamic.DYNAMIC in this case.

Reply:
Ahh good call